### PR TITLE
[Feature] TimeInput

### PR DESCRIPTION
**Pull request by riitasointi**
## Description

This PR introduces a new `<TimeInput>` component. It is very close to TextInput, but has a few autocomplete features on blur event. These instances are documented at the top of the component's Styleguidist page.

Also added a new small new feature to `<DateInput>` in this PR. This is logical because this new feature is needed when implementing the date + time pattern.  

## How Has This Been Tested?

Styleguidist. Chrome and Safari with VoiceOver

### TimeInput
* Introduce the TimeInput component

### DateInput
* Add `onDatePickerButtonBlur()` prop

### CODE REVIEW

File: src/core/Form/TimeInput/TimeInput.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
+The `<TimeInput>` component is used to input times. It looks and behaves similarly to a regular `<TextInput>` but has the following autocomplete features on input blur:
+
+- If the user types 1 or 2 characters which are valid hours (0-24), the time will be autocompleted. A leading zero will also be removed.
+  - E.g. 14 --> 14.00
+  - E.g. 09 --> 9.00 and 9 --> 9.00
+- If the user types 4 characters which form a valid "military time" value, the dot `.` character will be added as a separator. A leading zero will also be removed.
+  - E.g. 1745 --> 17.45 and 0200 --> 2.00
+- If the user types an otherwise valid time with a leading zero in the hours, the zero will be removed
+  - E.g. 07.45 --> 7.45
+- If the user types the colon `:` character it will be replaced with the dot `.` character. Dot is the correct time separator character in the Finnish language.
```
Comment (LJKaski) [inline src/core/Form/TimeInput/TimeInput.md:10]: I find the amount of 'E.g.'s a bit difficult to read. I'd try to simplify this list somehow. Maybe iterate the rules first and then provide example inputs and outputs. Even a table/chart might work 🤔

Comment (LJKaski) [reply src/core/Form/TimeInput/TimeInput.md:10]: These could even be a section with its own heading.

File: src/core/Form/TimeInput/TimeInput.tsx [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+   * not the blur FocusEvent like in other input components in this library!
+   */
+  onBlur?: (inputValue: string) => void;
+  /** Label for the input */
+  labelText: ReactNode;
+  /** Hides or shows the label. Label element is always present, but can be visually hidden.
+   * @default visible
+   */
+  labelMode?: LabelMode;
+  /** Hint text to be shown below the component */
```
Comment (LJKaski) [inline src/core/Form/TimeInput/TimeInput.tsx:74]: Should probably say 'between the label and the input' or something of the sort?

File: src/core/Form/TimeInput/TimeInput.tsx [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+  };
+
+  const handleOnBlur = () => {
+    let adjustedInputValue = '';
+    const inputValInt = parseInt(inputValue, 10);
+
+    // Handle automatic filling of 1 or 2 characters: 14 --> 14.00.
+    // Also remove leading zero from hours
+    if (
+      (inputValue.match(/^\d{1}$/) || inputValue.match(/^\d{2}$/)) &&
```
Comment (LJKaski) [inline src/core/Form/TimeInput/TimeInput.tsx:159]: Could this be simplified to just `(inputValue.match(/^\d{1,2}$/)`?

File: src/core/Form/TimeInput/TimeInput.tsx [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+  /** Controlled value */
+  value?: string;
+  /** Sets component's width to 100% of its parent */
+  fullWidth?: boolean;
+  /** Text to mark the field optional. Will be wrapped in parentheses and shown after `labelText` */
+  optionalText?: string;
+  /** Debounce time in milliseconds for onChange function. No debounce is applied if no value is given. */
+  debounce?: number;
+  /** Suomi.fi icon to be shown inside the input field */
+  icon?: ReactElement;
```
Comment (LJKaski) [inline src/core/Form/TimeInput/TimeInput.tsx:101]: Does the time input need icon support?

File: src/core/Form/TimeInput/TimeInput.md (hunk 1 above)
Comment (riitasointi) [reply src/core/Form/TimeInput/TimeInput.md]: I removed the repeating "e.g"s and adjusted the wording a bit. Made it easier to read IMO

File: src/core/Form/TimeInput/TimeInput.tsx (hunk 2 above)
Comment (riitasointi) [reply src/core/Form/TimeInput/TimeInput.tsx]: Yes, fixed!

File: src/core/Form/TimeInput/TimeInput.tsx (hunk 3 above)
Comment (riitasointi) [reply src/core/Form/TimeInput/TimeInput.tsx]: Indeed! Changed it now

File: src/core/Form/TimeInput/TimeInput.md (hunk 1 above)
Comment (LJKaski) [reply src/core/Form/TimeInput/TimeInput.md]: Looks good 👍

**Review by LJKaski — APPROVED** [review]

File: src/utils/common/common.ts [hunk 5]
```diff
@@ -47,6 +47,61 @@ export const forkRefs =
...
+ *  - 1-2 characters which form a valid hour designation are autocompleted to full time: 14 --> 14.00 and 9 --> 9.00
+ *  - 4 characters which form a valid "military time" will be autocompleted to full time: 1400 --> 14.00
+ *  - A leading zero will be removed from hours in an otherwise valid time: 09.00 --> 9.00
+ *  - The colon : character will be replaced by the dot . character in an otherwise valid time: 12:00 --> 12.00
+ * @param timeStr Time input value
+ * @returns An autocompleted time string or `null` if no automatic completions could be perfomed
+ */
+export const autocompleteTimeString = (timeStr: string): string | null => {
+  const inputValInt = parseInt(timeStr, 10);
+
```
Comment (LJKaski) [inline src/utils/common/common.ts:75]: Could there be a simple overall validity check in the beginning to avoid unnecessary if clause checks?

Comment (riitasointi) [reply src/utils/common/common.ts:77]: Added some now

**Review by LJKaski — DISMISSED** [review]

**Review by LJKaski — APPROVED** [review]

**Pull request merged**